search-app/src/cmr/search/services/content_service.clj

```diff
@@ -0,0 +1,150 @@
+  ([context route content]
+   (debug "Caching" route)
+   (cache/set-value (cache/context->cache context cache-key) route content))
+  ([context route provider-id tag]
+    (cache-page context
+                route
+                (get-page-content context route provider-id tag))))
+
+(defn- create-lookup-fn
+  [args]
+  (debug "Route" (second args) "not found in cache")
+  (fn [] (apply get-page-content args)))
+
+(defn retrieve-page
+  "Attempt to retrieve from the cache the page data for a given route. If no
+  value is found in the cache, genereate the content, cache it, and then return
+  the cached value.
+
+  Note that the first three arguements will always be `context`, HTTP `params`,
+  and resource `route`. Additional arguments may be present (e.g., provider id
+  and tag)."
+  [context params route & remaining]
+  (debug "Retrieving" route "with params" params "and remaining args" remaining)
+  (let [args (concat [context route] remaining)
+        regenerate? (= "true" (util/safe-lowercase (:regenerate params)))]
+    ;; Only admins can force the pages and sitemaps in the cache to be
+    ;; regenerated.
+    (when regenerate?
+      (acl/verify-ingest-management-permission context :update)
+      (debug "Forcing resource regeneration/cache update" route)
+      (apply cache-page args))
+    (cache/get-value (cache/context->cache context cache-key)
+                     route
+                     (create-lookup-fn args))))
+
+(defn- cache-top-level-content
+  "Cache all the content for the expensive, top-level routes."
+  [context]
+  (cache-page context "/sitemap.xml")
+  (cache-page context "/site/sitemap.xml")
+  (cache-page context "/site/collections/directory/eosdis"))
+
+(defn- cache-directory-content
+  "Cache all the content for the expensive, directory-level routes."
+  [context]
+  (debug "Providers:" (mdb/get-providers context))
+  (debug "Provider ids:" (map :provider-id (mdb/get-providers context)))
+  (doseq [provider-id (map :provider-id (mdb/get-providers context))]
+    (doseq [tag (keys site-utils/supported-directory-tags)]
+      (doseq [route [(format "/site/collections/directory/%s/%s" provider-id tag)
+                     (format "/site/collections/directory/%s/%s/sitemap.xml"
+                             provider-id
+                             tag)]]
+        (cache-page context route provider-id tag)))))
+
+(defn generate-content
+  "Cache all the content for the expensive routes. This is the function
+  intended for use by the static content job scheduler."
+  [context]
+  (info "Generating site content for caching")
+  (let [[ms _] (util/time-execution
+                (do
+                  (cache-top-level-content context)
+                  (cache-directory-content context)))]
+    (info (format "Content generated in %d ms" ms))))
+
+;; A job for generating static content
+;; Note: since content generation is being done via a job without an HTTP
```

Comment:
I'm worried that we're having to do something different here for just this job context. Why don't other jobs have the same problem?

Reply:
Dunno; I looked into it and couldn't find out why there were permissions errors when an HTTP request context wasn't present. I pinged others on-channel and no one else had any ideas. What's here is due to expediency.